extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeDashManifestCreator.java

```diff
@@ -1589,383 +1396,103 @@ private static void generateSegmentElementForPostLiveDvrStreams(
+        documentBuilderFactory.setExpandEntityReferences(false);
 
-    /**
-     * Set the limit of cached OTF manifests, cached post-live-DVR manifests and cached progressive
-     * manifests.
-     *
-     * <p>
-     * When the caches limit size are reached, oldest manifests will be removed from their
-     * respective cache.
-     * </p>
-     *
-     * <p>
-     * For each cache, if its new size set is less than the number of current cached manifests,
-     * oldest manifests will be also removed.
-     * </p>
-     *
-     * <p>
-     * Note that the limit must be more than 0 or an {@link IllegalArgumentException} will be
-     * thrown.
-     * </p>
-     *
-     * @param cachesLimit the maximum size of OTF, post-live-DVR and progressive caches
-     */
-    public static void setManifestsCachesMaximumSize(final int cachesLimit) {
-        GENERATED_OTF_MANIFESTS.setMaximumSize(cachesLimit);
-        GENERATED_POST_LIVE_DVR_STREAMS_MANIFESTS.setMaximumSize(cachesLimit);
-        GENERATED_PROGRESSIVE_STREAMS_MANIFESTS.setMaximumSize(cachesLimit);
+        final DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        return documentBuilder.newDocument();
     }
 
     /**
-     * Clear cached OTF manifests.
-     *
-     * <p>
-     * The limit of this cache size set, if there is one, will be not unset.
-     * </p>
+     * Securing against XEE is done by setting {@link XMLConstants#FEATURE_SECURE_PROCESSING} to
+     * {@code true} in the {@link TransformerFactory}, also see
+     * <a href="https://github.com/ChuckerTeam/chucker/pull/201">ChuckerTeam/chucker#201</a>.
+     * The best way to do this would be setting the attributes {@link
+     * XMLConstants#ACCESS_EXTERNAL_DTD} and {@link XMLConstants#ACCESS_EXTERNAL_STYLESHEET}, but
+     * unfortunately the engine on Android does not support them.
+     *
+     * @param document the document to convert; must have been created using {@link #newDocument()}
+     *                 to properly prevent XEE attacks!
+     * @return the document converted to an XML string, making sure there can't be XEE attacks
      */
-    public static void clearOtfCachedManifests() {
-        GENERATED_OTF_MANIFESTS.clear();
-    }
+    private static String documentToXml(@Nonnull final Document document)
+            throws TransformerException {
 
-    /**
-     * Clear cached post-live-DVR streams manifests.
-     *
-     * <p>
-     * The limit of this cache size set, if there is one, will be not unset.
-     * </p>
-     */
-    public static void clearPostLiveDvrStreamsCachedManifests() {
-        GENERATED_POST_LIVE_DVR_STREAMS_MANIFESTS.clear();
-    }
+        @SuppressWarnings("java:S2755") // see javadoc: this is actually taken care of
+        final TransformerFactory transformerFactory = TransformerFactory.newInstance();
+        transformerFactory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
```

Comment:
1. It's called XXE
2. About that "unfortunately the engine on Android does not support them":


Android - as nearly always - uses rubbish stuff. They have a custom implementation
![grafik](https://user-images.githubusercontent.com/40789489/167313203-2b8b20f2-a8cf-4ae6-943a-b7cb1dabd674.png)
https://android.googlesource.com/platform/external/apache-xml/+/refs/heads/master/src/main/java/org/apache/xalan/processor/TransformerFactoryImpl.java (based on Apache Xalan)
because why follow the standard...

On the other hand: You recognized that this is a Java project or? so we code against the Java API - This "fix" still leave a backdoor open on Java:
* https://docs.oracle.com/javase/7/docs/api/javax/xml/XMLConstants.html#ACCESS_EXTERNAL_SCHEMA
  > When FEATURE_SECURE_PROCESSING is enabled, it is **recommended** that implementations restrict external connections by default...
  
  The problem here is recommend != it's actual the case.

TL;DR
I recommend doing the following:
Add the fixes from https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html#transformerfactory / https://rules.sonarsource.com/java/RSPEC-2755 and put a try-catch block around them.
If an exception occurs (e.g. on Android) we simply ignore it. If not well - we are now definitely safe.

PS: XML (or at least Java's implementation) is evil/extremely badly implemented, don't use it if you can - it will sooner or later cause security holes. Use JSON or similar instead.

Reply:
Done